### ISSUE TITLE
Idempotent Producer support

### ISSUE THREAD

**Issue by bshafiee** [confluentinc/confluent-kafka-go#238]
Description
===========
Hello, 
 I was just wondering when Idempotent Producer is gonna be available in the Go library. I see it's merged in the `C` repo. I did try it by enabling it in the config(`		"enable.idempotence":        "true"`), and it did work until some errors happened (similar to the one described here: https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#producer-message-delivery-failure):
```
ProduceRequest with 1 message(s) succeeded when expecting failure (broker 43 PID{Id:601000,Epoch:0}, base seq 43821, next ack seq 43820, next err seq 43821: unable to retry without risking duplication/reordering
``` 

and the producer stopped producing anything. Then I was unable to find any API in the Go lib to fetch the persistence status and wasn't sure how to deal with those errors so the producer can move on.  It would be really useful if you could provide some example of Idempotent Producer usage and how to deal when errors happen so the application doesn't get stuck. 

For now I disabled `enable.gapless.guarantee` since I don't really care about individual messages and would like the producer to move on, but not sure if this is the right approach!

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-go and librdkafka version (`LibraryVersion()`):
 - [ ] Apache Kafka broker version:
 - [ ] Client configuration: `ConfigMap{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Fatal errors are being added in #240

**Comment by bshafiee** [comment]
Awesome, let me know if I can help with the Go lib!
So in my case, IIUC since the producer is in a non-recoverable state, its PID should be reset to continue. Now can this be achieved somehow without re-instantiating the producer object (`prouder = kafka.NewProcuer...`)?  This kinda sucks since then I will need to have some mutex for using `producer.*` in other go routines and the thread-safety of producer will be compromised. Would be possible to add a method to just reset PID?

**Comment by edenhill** [comment]
That is not currently possible.

What would need to happen is purge and fail all messages in queue, so that a delivery report is emitted for each one, and then allow the application re-initialize the pid prior to any messages have been produce()d.

But, if the application sends the same message again it might lead to duplicates and/or out of sequence, which is what idempotence is out to protect against.

What would your strategy be in this case?

**Comment by bshafiee** [comment]
Correct, I am on-board with draining the items left in the queue and failing each one of them. What I was saying was maybe we can do this internally as opposed to re-instantiate the producer object. 

>if the application sends the same message again it might lead to duplicates and/or out of sequence, which is what idempotence is out to protect against.

I thought that was part of the protocol to skip the already delivered ones. 

Regardless, the reason I wanted to switch to idempotent producer was to reduce duplicates not completely provide `exactly once`, currently about 1-2% of my produced messages are duplicates to achieve `at least once` guarantee.

**Comment by edenhill** [comment]
> I thought that was part of the protocol to skip the already delivered ones.

It is, within a producer session with the same PID.
The producer can't keep track of messages that are outside its control (such as messages failed and re-produce()d by the application).

You could disable the gapless guarantee to be less stringent for some type of errors, see here:
https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#guarantees

**Comment by edenhill** [comment]
I think "ProduceRequest with 1 message(s) succeeded when expecting failure " is a bug, treating a previous error incorrectly.
I'm adding more info to that log message to help troubleshoot it further.

Do you know how to reproduce it?

**Comment by bshafiee** [comment]
I think it happens every couple of hours (I am not sure how to reproduce it) but I can certainly add more logs and report it back when it happens again. Let me know when you pushed your changes for more logging. 
If I were to guess, I think it happens when a leader election happens or something changes with brokers (I don't really control our brokers so hard to tell). 

But regardless, if we hit such a state, what does a client really can do? except re-instantiating and prolly reproduce the same or just move on (unless some fancy offset tracking logic exist in which case why would use idempotent producer in the first place). So I think maybe it's fair to provide such a method that drains and resets the PID as a convenient method (with stating the consequences).

**Comment by edenhill** [comment]
Thank you, I've reproduced this.
This is a bug that is triggered when a previous request times out but subsequent requests succeed.
I have a fix on the way that treats this as an implicit ack of the previous requests (since the subsequent request would fail if the first ones did not make it).

**Comment by edenhill** [comment]
Idempotent Producer is available in the v1.0.0 release.
